Add duplication command

wiktor-k opened the pull request:
Hello :wave: 

I'm working on TPM duplication/import feature and would like to gradually propose changes if you don't mind.

The current code compiles but the test fails with `Error when performing duplication: authorization handle is not correct for command`. Any suggestions welcome... and I'll debug this in parallel.

Edit: this is still raw but I'm updating the PR as I go...

ionut-arm commented on tss-esapi/tests/context_tests/tpm_commands/duplication_commands_tests.rs:
```diff
@@ -1,2 +1,57 @@
+            .unwrap();
+        context.tr_set_auth(new_parent_handle, &key_auth).unwrap();
+
+        let _ = context
+            .duplicate(
+                object_to_duplicate_handle,
+                new_parent_handle,
+                None,
+                SymmetricDefinitionObject::Null,
+            )
+            .unwrap();
```
I think you need to crate a policy session for this call and follow the steps mentioned in the spec for this command to set it up. Unfortunately, it's probably going to be somewhat trial-and-error.

wiktor-k replied:
Okay, thanks for the hint! I already had a policy session that allowed duplication but then key creation failed. I'll check it out, thanks!

ionut-arm replied:
I think that session should only be used for the duplicate call, not for creation as well!

ionut-arm commented on tss-esapi/src/context/tpm_commands/duplication_commands.rs:
```diff
@@ -1,9 +1,57 @@
+                self.mut_context(),
+                object_to_duplicate.into(),
+                new_parent_handle.into(),
+                self.optional_session_1(),
```
This session won't be optional, so `self.required_session_1()`

Superhepper commented on tss-esapi/tests/context_tests/tpm_commands/duplication_commands_tests.rs:
```diff
@@ -1,2 +1,131 @@
+        let random_digest = context.get_random(16).unwrap();
+        let key_auth = Auth::try_from(random_digest.value().to_vec()).unwrap();
+
+        let object_to_duplicate_handle: ObjectHandle = context
```
I do not think you can duplicate a Primary key. Because duplicate is used to move an object from one Hierarchy to another (potentially on another machine)

wiktor-k commented:
Okay, phew. Finally got it working by basically memorizing the integration test from here: https://github.com/utkarsh2102/tpm2-tss/blob/5eca0a4ae129f410303de1c20046ebe7527b9ad3/test/integration/esys-duplicate.int.c#L382 and rewriting it in Rust and our API.

This is ready for review now. One thing I considered is removing the separate `duplication_tests` since the doctest is basically testing the same... Opinions welcome!

Superhepper commented on tss-esapi/src/context/tpm_commands/duplication_commands.rs:
````diff
@@ -1,9 +1,334 @@
+    /// * `out_sym_seed` - Seed protected by the asymmetric algorithms of new parent.
+    ///
+    /// ```rust
+    /// use std::convert::{TryFrom, TryInto};
````
Really nice example!
It is nice that it runs and works as test. Though this will show up as an example in the documentation and there it will probably be way to long. So what we have done previously is to hide everything that does not directly involve calling the api that you are documenting. So that we only show the important parts but still the documentation will fail to compile if we change something (if you need some inspiration look at evict_control, that have like 300 lines of code but only about 20 is visible).

wiktor-k replied:
Understood. I was thinking about reducing the visible lines but there are quite some things that need to be set (like fixedTPM and fixedParent = false) for the example to work... I'll trim it down.

Superhepper commented on tss-esapi/tests/context_tests/tpm_commands/duplication_commands_tests.rs:
```diff
@@ -1,2 +1,193 @@
+            .with_rsa_parameters(
+                PublicRsaParametersBuilder::new()
+                    .with_scheme(RsaScheme::Null)
+                    .with_key_bits(RsaKeyBits::Rsa2048)
```
If you want your test to be different from your documentation then one way of doing it is to use another type of key like a ECC key.

wiktor-k replied:
Excellent idea! I'll rework the test case to use ECC keys.

wiktor-k commented:
Okay, improved the PR in a couple of ways: as suggested by Jesper: minimized the visible example code and changed the test case to use ECC. The latter proved to be more daunting than I anticipated as it was not clear which parameter combinations were invalid so I added more error reporting (the change is minimal so I didn't open a new PR). As usual: I'm open to suggestions w.r.t. error messages.

Thanks for your time! :bow:

ionut-arm reviewed (approved):
👍🏻 LGTM

wiktor-k commented:
Greatly appreciated @ionut-arm ! :bow:
